File: src/echo-actors/src/main/scala/echo/actor/directory/service/FeedService.scala
package echo.actor.directory.service

import javax.persistence.EntityManagerFactory
import javax.transaction.Transactional

import echo.actor.directory.orm.FeedDao
import echo.actor.directory.orm.impl.FeedDaoImpl
import echo.actor.directory.repository.{FeedRepository, RepositoryFactoryBuilder}
import echo.core.converter.mapper.FeedMapper
import echo.core.model.dto.FeedDTO
import org.springframework.orm.jpa.EntityManagerHolder
import org.springframework.transaction.support.TransactionSynchronizationManager

/**
  * @author Maximilian Irro
  */
class FeedService(private val repositoryFactoryBuilder: RepositoryFactoryBuilder) {

    private val repositoryFactory = repositoryFactoryBuilder.createFactory
    private val feedRepository: FeedRepository = repositoryFactory.getRepository(classOf[FeedRepository])

    private val em = repositoryFactoryBuilder.getEntityManager
    private val emf: EntityManagerFactory = repositoryFactoryBuilder.getEntityManagerFactory
    private val feedDao: FeedDao =  new FeedDaoImpl(emf)

    @Transactional
    def save(feedDTO: FeedDTO): FeedDTO = {
        //val em = repositoryFactoryBuilder.getEntityManager
        //val emf = repositoryFactoryBuilder.getEntityManagerFactory

        /*
        val tx = em.getTransaction
        feedRepository.save(feed)
        tx.commit
        */

        /*
        TransactionSynchronizationManager.bindResource(emf, new EntityManagerHolder(em))
        try {
            val feed = FeedMapper.INSTANCE.feedDtoToFeed(feedDTO)
            val result = feedRepository.save(feed)
            FeedMapper.INSTANCE.feedToFeedDto(result)
        } finally {
            // Make sure to unbind when done with the repository instance
            TransactionSynchronizationManager.unbindResource(emf)
        }
        */


        val feed = FeedMapper.INSTANCE.feedDtoToFeed(feedDTO)
        val result = feedRepository.save(feed)
        FeedMapper.INSTANCE.feedToFeedDto(result)

    }

    @Transactional
    def findOne(id: Long): FeedDTO = {
        val result = feedRepository.findOne(id)
        FeedMapper.INSTANCE.feedToFeedDto(result)
    }

    @Transactional
    def findOneByUrl(url: String): FeedDTO = {
        val result = feedRepository.findOneByUrl(url)
        FeedMapper.INSTANCE.feedToFeedDto(result)
    }

}

File: src/echo-microservices/app-directory/src/main/java/echo/microservice/directory/service/FeedService.java
package echo.microservice.directory.service;

import echo.core.domain.dto.FeedDTO;
import echo.microservice.directory.repository.FeedRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

@Service
@Transactional
public class FeedService {

    @Autowired
    private FeedRepository feedRepository;

    @Transactional(readOnly = true)
    public Optional<FeedDTO> findOne(Long id) {
        throw new UnsupportedOperationException("Not yet implemented"); // TODO
    }

    @Transactional(readOnly = true)
    public Optional<FeedDTO> findOneByEchoId(String echoId) {
        throw new UnsupportedOperationException("Not yet implemented"); // TODO
    }

}
